### PR TITLE
Improve fault-handling based on received fault-message

### PR DESCRIPTION
**Pull request by bouwew**
While developing and improving the Waqi-test custom_component (https://github.com/bouwew/waqi-client-async) based on your earlier work, I noticed that the error-handling did not work very well. I would mostly receive Unknow Error when inducing a fault.

This PR implements some of the improvements that can be made based on my testing.
I have also added a debug-logging line to be able to see the raw data being received from the Waqi-API, so we are able to see exactly what is returned.

### CODE REVIEW

**Comment by sbach** [comment]
Thank you for the PR @bouwew!

I agree that the error handling is something I had kept to the bare minimum. I appreciate you improving on this!

Before I merge, can you please run `poetry run black waqi_client_async/` on your commits?

FYI: I readied a GitHub Actions workflow to help with validation.

**Comment by bouwew** [comment]
Sure no problem, done.

File: waqi_client_async/__init__.py [hunk 1]
```diff
@@ -29,30 +33,51 @@ class OverQuota(APIError):
...
-        return
+    LOGGER.debug("JSON Data: %s", result)
+    if result.get("status") == "ok":
+        if data := result.get("data") is not None:
+            if data.get("msg") == "Unknown ID":
+                raise UnknownID()
 
-    data = result.get("data")
-    if data == "Invalid key":
+    return
```
Comment (sbach) [inline waqi_client_async/__init__.py]: This means the code below does not get executed. What should we do here?

File: waqi_client_async/__init__.py [hunk 2]
```diff
@@ -29,30 +33,51 @@ class OverQuota(APIError):
...
 
     pass
 
 
 def assert_valid(result: dict) -> None:
-    if result.get("status") != "error":
-        return
+    LOGGER.debug("JSON Data: %s", result)
+    if result.get("status") == "ok":
+        if data := result.get("data") is not None:
```
Comment (sbach) [inline waqi_client_async/__init__.py]: This `:=` is a Python 3.8 feature. This project was configured for Python >=3.7 (see [here](https://github.com/sbach/waqi-client-async/blob/main/pyproject.toml#L15))

Comment (sbach) [reply waqi_client_async/__init__.py]: The question becomes: what is our intent here?
- Do we update the minimal Python version supported?
- Do we update the code to be compatible?

File: waqi_client_async/__init__.py [hunk 3]
```diff
@@ -29,30 +33,51 @@ class OverQuota(APIError):
...
         raise APIError(message)
 
+    keyword = result.get("keyword")
+    if keyword == "no_matching_stations_found":
+        raise APIError(keyword)
+
+    data = result.get("data")
+    if data == "no such station":
+        raise UnknownStation()
+
```
Comment (sbach) [inline waqi_client_async/__init__.py]: Enabling this code by removing the above `return`, I have been trying, without success, to reproduce the potential JSON result that would match those updated checks. Could you please share the URLs for which it reproduced? I would want to write `pytest` tests for that.

**Review by sbach — CHANGES_REQUESTED** [review]
I took a serious look at your changes today. Could you help me address those comments?

File: waqi_client_async/__init__.py (hunk 1 above)
Comment (bouwew) [reply waqi_client_async/__init__.py]: Fixed, needs one more ident.

File: waqi_client_async/__init__.py (hunk 2 above)
Comment (bouwew) [reply waqi_client_async/__init__.py]: HA is now on python 3.11, so I'd propose raising the minimal supported version.

File: waqi_client_async/__init__.py (hunk 3 above)
Comment (bouwew) [reply waqi_client_async/__init__.py]: I have captured some results that has lead me to write these code-snippets.

But, as I can't see the full data-packets that are returned, I'm not sure if they are correct.

I have added a `LOGGER.debug`-line with the purpose to be able to see the full data-packets that are returned.
With this complete info I had planned to make the error-handling code final and fully working.

So, please release a version 1.0.1 based on the present, corrected code (see my latest commits below) and allow me to capture more data, that I happily will share with you. And let's then create a v1.1.0 with correct error-handling and added test-cases.

File: waqi_client_async/__init__.py (hunk 2 above)
Comment (bouwew) [reply waqi_client_async/__init__.py]: And implemented.

**Comment by bouwew** [comment]
FYI, found yet another waqi-related module: https://github.com/Ozon3Org/Ozon3

**Comment by bouwew** [comment]
About my request to release v1.0.1, never mind, I've managed to release to testpypi.

**Comment by bouwew** [comment]
Ok, I'll perform some more testing now that I have managed to publish to Testpypi.

**Comment by bouwew** [comment]
Ok, got it working via Testpypi but I needed to go back to your original assert_valid() code due to some strange exception error.
This is what I now receive as output:
```
Waqi failed station search:
2023-03-18 11:33:55.490 DEBUG (MainThread) [waqi_client_async] JSON Data: {'status': 'ok', 'data': []}

WAQI station search with bad API_TOKEN (0000):
2023-03-18 11:52:04.714 DEBUG (MainThread) [waqi_client_async] JSON Data: {'status': 'error', 'data': 'Invalid key'}

Waqi feed search with A113074:
2023-03-18 11:49:37.062 DEBUG (MainThread) [waqi_client_async] JSON Data: {'status': 'ok', 'data': {'aqi': 6, 'idx': -113074, 'attributions': [{'url': 'https://luftdaten.info', 'name': 'Citizen Science project luftdaten.info', 'station': '44536'}, {'url': 'https://waqi.info/', 'name': 'World Air Quality Index Project'}], 'city': {'geo': [51.966, 6.292], 'name': 'Doctor Huber Noodtplaats', 'url': 'https://aqicn.org/station/@113074', 'location': 'Doctor Huber Noodtplaats, Doetinchem, Gelderland, Netherlands, 7001DT, The Netherlands'}, 'dominentpol': 'pm25', 'iaqi': {'pm10': {'v': 3}, 'pm25': {'v': 6}}, 'time': {'s': '2023-03-18 11:43:01', 'tz': '+01:00', 'v': 1679136181, 'iso': '2023-03-18T10:43:01Z'}}}

Waqi failed feed search, using malformed feed id: @113074
2023-03-18 11:40:48.468 DEBUG (MainThread) [waqi_client_async] JSON Data: {'status': 'ok', 'data': {'status': 'error', 'msg': 'Unknown ID'}}

Waqi feed search wit bad API_TOKEN (0000):
2023-03-18 11:45:08.614 DEBUG (MainThread) [waqi_client_async] JSON Data: {'dt': '205.204µs', 'rxs': {'obs': [{'msg': 'Invalid key', 'status': 'error'}], 'status': 'ok', 'ver': '1'}}
```

**Comment by bouwew** [comment]
The first two bits of data are probably familiar to you.

The other bits of data show what comes back from the feed-search option that I've added.

**Comment by bouwew** [comment]
Looks like my latest code is working :) please see the above updates.

Also the custom_component code was updated: https://github.com/bouwew/waqi-test

**Comment by bouwew** [comment]
Various more updates. I also fixed the testing, now the actual code that's in the development-branch will be tested instead of the released version on Pypi (import takes the version that's on Pypi).

**Pull request closed without merging**